File: recursion/combination_sum_ii.cpp
//{ Driver Code Starts
// Initial Template for C++

#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
// User function Template for C++

class Solution {
  vector<vector<int>> ans;

  void solve(const vector<int> &a, int i, int n, int k, vector<int> &curr) {
    if (k == 0) {
      if (ans.size() == 0 || ans.back() != curr)
        ans.push_back(curr);
      return;
    }
    if (i == n || k < 0) {
      return;
    }
    curr.push_back(a[i]);
    solve(a, i + 1, n, k - a[i], curr);
    curr.pop_back();
    int j = i + 1;
    while (j < n && a[j] == a[i])
      ++j;
    solve(a, j, n, k, curr);
  }

public:
  vector<vector<int>> CombinationSum2(vector<int> arr, int n, int k) {
    // code here
    ans.clear();
    sort(arr.begin(), arr.end());
    vector<int> curr;
    solve(arr, 0, n, k, curr);
    return ans;
  }
};

//{ Driver Code Starts.

int main() {
  int t;
  cin >> t;
  while (t--) {
    int n, k;
    cin >> n >> k;
    vector<int> arr(n);
    for (int i = 0; i < n; i++) {
      cin >> arr[i];
    }
    Solution ob;
    auto obj = ob.CombinationSum2(arr, n, k);
    for (int i = 0; i < obj.size(); i++) {
      for (int j = 0; j < obj[i].size(); j++) {
        cout << obj[i][j] << " ";
      }
      cout << "\n";
    }
    if (obj.size() == 0)
      cout << "\n";
  }
  return 0;
}
// } Driver Code Ends

File: recursion/combination_sum_ii.py
# User function Template for python3


class Solution:
    def solve(self, c, k, t, curr):
        if t == 0:
            self.ans.append(curr.copy())
            return
        if t < 0 or k == n:
            return

        if t < c[k]:
            return
        curr.append(c[k])
        self.solve(c, k + 1, t - c[k], curr)
        curr.pop()
        j = k + 1
        while j < n and c[j] == c[k]:
            j += 1
        self.solve(c, j, t, curr)

    def combinationSum2(self, candidates, target):
        # Code here
        candidates.sort()
        self.ans = []
        self.solve(candidates, 0, target, [])
        return self.ans


# {
# Driver Code Starts
# Initial Template for Python 3

if __name__ == "__main__":
    t = int(input())
    for _ in range(t):
        n, target = list(map(int, input().split()))
        candidates = list(map(int, input().split()))
        ob = Solution()
        res = ob.combinationSum2(candidates, target)
        res.sort()
        print("[ ", end="")
        for subset in res:
            print("[ ", end="")
            for val in subset:
                print(val, end=" ")
            print("]", end="")
        print(" ]")
# } Driver Code Ends
